Crash when method precondition quantifies over "provided" type in export

The following program causes Dafny to crash during translation into boogie:

```
module M1 {
  export Abs provides m, T
  export Conc provides m, MyClass reveals T

  class MyClass {

  }

  type T = MyClass

  lemma m(f : T ~> bool)
    requires forall t :: f.requires(t) ==> f(t)
    { }

}

module M2 {
  import M1`Abs

  method k(t : M1.T){

  }
}
```

With the error message

```
[ERROR] FATAL UNHANDLED EXCEPTION: System.Diagnostics.Contracts.ContractException: Assertion failed.
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailureImplementation (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0002f] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Runtime.CompilerServices.ContractHelper.TriggerFailure (System.Diagnostics.Contracts.ContractFailureKind kind, System.String displayMessage, System.String userMessage, System.String conditionText, System.Exception innerException) [0x00000] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Diagnostics.Contracts.Contract.ReportFailure (System.Diagnostics.Contracts.ContractFailureKind failureKind, System.String userMessage, System.String conditionText, System.Exception innerException) [0x0003a] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Diagnostics.Contracts.Contract.Assert (System.Boolean condition) [0x00003] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at Microsoft.Dafny.Type.NormalizeExpand (System.Boolean keepConstraints) [0x00083] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator.TrType (Microsoft.Dafny.Type type) [0x00004] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator+ExpressionTranslator.TrBoundVariables (System.Collections.Generic.List`1[T] boundVars, System.Collections.Generic.List`1[T] bvars, System.Boolean translateAsLocals, System.Collections.Generic.List`1[T] freeOfAlloc) [0x00047] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator+ExpressionTranslator.TrExpr (Microsoft.Dafny.Expression expr) [0x047dd] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator.TrSplitExpr (Microsoft.Dafny.Expression expr, System.Collections.Generic.List`1[T] splits, System.Boolean position, System.Int32 heightLimit, System.Boolean inlineProtectedFunctions, System.Boolean apply_induction, Microsoft.Dafny.Translator+ExpressionTranslator etran) [0x015aa] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator.TrSplitExpr (Microsoft.Dafny.Expression expr, System.Collections.Generic.List`1[T] splits, System.Boolean position, System.Int32 heightLimit, System.Boolean inlineProtectedFunctions, System.Boolean apply_induction, Microsoft.Dafny.Translator+ExpressionTranslator etran) [0x010ab] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator.TrSplitExprForMethodSpec (Microsoft.Dafny.Expression expr, Microsoft.Dafny.Translator+ExpressionTranslator etran, Microsoft.Dafny.Translator+MethodTranslationKind kind) [0x0000a] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator.AddMethod (Microsoft.Dafny.Method m, Microsoft.Dafny.Translator+MethodTranslationKind kind) [0x003c9] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator.AddMethod_Top (Microsoft.Dafny.Method m) [0x000aa] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator.AddClassMembers (Microsoft.Dafny.TopLevelDeclWithMembers c, System.Boolean includeMethods) [0x003b3] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator.AddTypeDecl (Microsoft.Dafny.RevealableTypeDecl d) [0x00138] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator.DoTranslation (Microsoft.Dafny.Program p, Microsoft.Dafny.ModuleDefinition forModule) [0x00437] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.Translator+<Translate>d__47.MoveNext () [0x000ea] in <34d3a135e58a44ba97fa223128f8d916>:0
  at Microsoft.Dafny.DafnyDriver+<Translate>d__6.MoveNext () [0x00119] in <cd1f2aa755c64f5bbfa0950a1869826c>:0
  at Microsoft.Dafny.DafnyDriver.Boogie (System.String baseName, System.Collections.Generic.IEnumerable`1[T] boogiePrograms, System.String programId, System.Collections.Generic.Dictionary`2[System.String,Microsoft.Boogie.PipelineStatistics]& statss, Microsoft.Boogie.PipelineOutcome& oc) [0x0012f] in <cd1f2aa755c64f5bbfa0950a1869826c>:0
  at Microsoft.Dafny.DafnyDriver.ProcessFiles (System.Collections.Generic.IList`1[T] dafnyFiles, System.Collections.ObjectModel.ReadOnlyCollection`1[T] otherFileNames, Microsoft.Dafny.ErrorReporter reporter, System.Boolean lookForSnapshots, System.String programId) [0x00222] in <cd1f2aa755c64f5bbfa0950a1869826c>:0
  at Microsoft.Dafny.DafnyDriver.ThreadMain (System.String[] args) [0x0003a] in <cd1f2aa755c64f5bbfa0950a1869826c>:0
  at Microsoft.Dafny.DafnyDriver+<>c__DisplayClass1_0.<Main>b__0 () [0x00001] in <cd1f2aa755c64f5bbfa0950a1869826c>:0
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) [0x00014] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) [0x0002b] in <f9d1b832704f410aa8ec771f4fe80552>:0
  at System.Threading.ThreadHelper.ThreadStart () [0x00008] in <f9d1b832704f410aa8ec771f4fe80552>:0
```

If `M1` is instead imported as `Conc` (i.e. revealing that `T` is an alias to `MyClass`) then it works as expected.

--- a/Source/Dafny/DafnyAst.cs
+++ b/Source/Dafny/DafnyAst.cs
@@ -1842,6 +1842,11 @@ namespace Microsoft.Dafny {
       Contract.Requires(b != null);
       Contract.Requires(builtIns != null);
 
+      // As a special-case optimization, check for equality here, which will better preserve un-expanded type synonyms
+      if (a.Equals(b, true)) {
+        return a;
+      }
+
       // Before we do anything else, make a note of whether or not both "a" and "b" are non-null types.
       var abNonNullTypes = a.IsNonNullRefType && b.IsNonNullRefType;
 

--- a/Source/Dafny/Resolver.cs
+++ b/Source/Dafny/Resolver.cs
@@ -4378,7 +4378,7 @@ namespace Microsoft.Dafny
       Contract.Requires(!(t is TypeProxy));
       Contract.Requires(!(t is ArtificialType));
 
-      t = keepConstraints ? t.NormalizeExpandKeepConstraints() : t.NormalizeExpand();
+      t = keepConstraints ? t.Normalize() : t.NormalizeExpand();
       // never violate the type constraint of a literal expression
       var followedRequestedAssignment = true;
       foreach (var su in proxy.Supertypes) {
@@ -6051,7 +6051,7 @@ namespace Microsoft.Dafny
       var meets = new List<Type>();
       foreach (var xc in AllXConstraints) {
         if (xc.ConstraintName == "Assignable" && xc.Types[1].Normalize() == proxy) {
-          var su = xc.Types[0].NormalizeExpandKeepConstraints();
+          var su = xc.Types[0].Normalize();
           if (su is TypeProxy) {
             continue;  // don't include proxies in the meet computation
           }

--- a/Test/git-issues/git-issue-623.dfy
+++ b/Test/git-issues/git-issue-623.dfy
@@ -1,0 +1,107 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+// ----- example reported in Issue 623
+
+module M1 {
+  export Abs
+    provides M, T
+  export Conc
+    provides M, MyClass
+    reveals T
+
+  class MyClass {
+  }
+
+  type T = MyClass
+
+  lemma M(f: T ~> bool)
+    requires forall t :: f.requires(t) ==> f(t)  // regression test: this once crashed during checking of M2
+  { }
+}
+
+module M2 {
+  import M1`Abs
+
+  method K(t: M1.T) {
+  }
+}
+
+module M3 {
+  import M1`Conc
+
+  method K(t: M1.T) {
+  }
+}
+
+// ----- example reported in Issue 150
+
+module N1 {
+  export
+    provides T, Equal, Foo
+
+  type T(==) = seq<real>
+
+  predicate Equal(u: T, v: T)
+  {
+    u == v
+  }
+
+  lemma Foo()
+    ensures forall u, v :: Equal(u, v) ==> u == v  // regression test: this once crashed during checking of N2
+  { }
+}
+
+module N2 {
+  import N1
+
+  lemma Bar(u: N1.T, v: N1.T)
+    requires N1.Equal(u, v)
+  {
+    N1.Foo();
+    assert u == v;
+  }
+}
+
+
+// ------------------- additional examples
+
+module Library {
+  export
+    provides W, P, X, Q
+    provides M0, M1
+
+  type W = MyTrait
+  trait MyTrait {
+  }
+  predicate P(u: W)
+
+  type X = MyClass
+  class MyClass extends MyTrait {
+  }
+  predicate Q(x: X)
+
+  lemma M0()
+    requires forall t :: P(t)
+  { }
+  lemma M1()
+    requires forall t :: Q(t)
+  { }
+
+  lemma Private0()
+    requires forall t :: P(t) && Q(t)  // error: t is inferred as MyTrait, so can't prove that Q(t) is well-formed
+  { }
+  lemma Private1()
+    requires forall t :: Q(t) && P(t)  // error: t is inferred as MyTrait, so can't prove that Q(t) is well-formed
+  { }
+}
+
+module Client {
+  import Library
+
+  method K()
+    requires forall t :: Library.P(t)
+    requires forall t :: Library.Q(t)
+  {
+  }
+}

--- a/Test/git-issues/git-issue-623.dfy.expect
+++ b/Test/git-issues/git-issue-623.dfy.expect
@@ -1,0 +1,8 @@
+git-issue-623.dfy(92,35): Error: the RHS value (a MyTrait) is not known to be an instance of the LHS type (MyClass)
+Execution trace:
+    (0,0): anon0
+git-issue-623.dfy(95,27): Error: the RHS value (a MyTrait) is not known to be an instance of the LHS type (MyClass)
+Execution trace:
+    (0,0): anon0
+
+Dafny program verifier finished with 3 verified, 2 errors

